zerver/lib/test_runner.py

```diff
@@ -491,9 +500,36 @@ def run_tests(self, test_labels: List[str],
         result = self.run_suite(suite)
         self.teardown_test_environment()
         failed = self.suite_result(suite, result)
-        if not failed:
+        failed_thread = []  # type: List[str]
+        if not failed or self.parallel == 1:
             write_instrumentation_reports(full_suite=full_suite, include_webhooks=include_webhooks)
-        return failed, result.failed_tests
+            return failed, result.failed_tests, failed_thread
+
+        # Below is the logic for passing the list of tests ran in a failed
+        # worker, for use in the --replicate option when running test-backend.
+        args = [
+            (index, subsuite[1])
+            for index, subsuite in enumerate(suite.subsuites)
+        ]  # type: List[Tuple[int, List[str]]]
+
+        # We iterate through the subsuites for the failure. When
+        # we find it, assign the index for use in aggregating the
+        # individual tests.
+        index = None
+        for indexed_tests in args:
+            for test in indexed_tests[1]:
+                if result.failed_tests[0] == test:
+                    index = indexed_tests[0]
+                    break
+        assert index is not None
+        failed_pid = _subsuite_index_to_pid[index]
+        failed_thread = [
+            test
+            for subsuite, pid in zip(suite.subsuites, _subsuite_index_to_pid)
+            if pid == failed_pid
+            for test in subsuite[1]
+        ]
+        return failed, result.failed_tests, failed_thread
```

Comment:
I think `failed_thread` should be called `failed_thread_tests`, since the thing is a list of tests, not a "thread".

Reply:
Sounds great, there were some inconsistencies where it was named `failed_thread_tests` in some places and not in others.  I addressed that as well.